src/core/algorithms/order/order.cpp

```diff
@@ -0,0 +1,390 @@
+                return mixed_type->CompareAsStrings(l.second, r.second) ==
+                       model::CompareResult::kLess;
+            }
+            return type->Compare(l.second, r.second) == model::CompareResult::kLess;
+        };
+        auto equal = [&type, &mixed_type](std::pair<unsigned long, std::byte const*> const& l,
+                                          std::pair<unsigned long, std::byte const*> const& r) {
+            if (type->GetTypeId() == +(model::TypeId::kMixed)) {
+                return mixed_type->CompareAsStrings(l.second, r.second) ==
+                       model::CompareResult::kEqual;
+            }
+            return type->Compare(l.second, r.second) == model::CompareResult::kEqual;
+        };
+        std::sort(indexed_byte_data.begin(), indexed_byte_data.end(), less);
+        SortedPartition::EquivalenceClasses equivalence_classes;
+        equivalence_classes.reserve(typed_relation_->GetNumRows());
+        equivalence_classes.push_back({indexed_byte_data.front().first});
+        for (size_t k = 1; k < indexed_byte_data.size(); ++k) {
+            if (equal(indexed_byte_data[k - 1], indexed_byte_data[k])) {
+                equivalence_classes.back().insert(indexed_byte_data[k].first);
+            } else {
+                equivalence_classes.push_back({indexed_byte_data[k].first});
+            }
+        }
+        equivalence_classes.shrink_to_fit();
+        sorted_partitions_.emplace(AttributeList{i}, SortedPartition(std::move(equivalence_classes),
+                                                            typed_relation_->GetNumRows()));
+    }
+    for (auto& [attr, partition] : sorted_partitions_) {
+        if (partition.Size() == 1) {
+            for (AttributeList other_attr : single_attributes_) {
+                if (other_attr != attr) {
+                    valid_[attr].insert(other_attr);
+                }
+            }
+            single_attributes_.erase(std::find(single_attributes_.begin(), single_attributes_.end(), attr));
+        }
+    }
+    for (auto& [attr, partition] : sorted_partitions_) {
+        if (partition.Size() == 1) {
+            for (AttributeList other_attr : single_attributes_) {
+                if (other_attr != attr) {
+                    valid_[attr].insert(other_attr);
+                }
+            }
+            single_attributes_.erase(std::find(single_attributes_.begin(), single_attributes_.end(), attr));
+        }
+    }
+}
+
+void Order::CreateSortedPartitionsFromSingletons(AttributeList const& attr_list) {
+    if (sorted_partitions_.find(attr_list) != sorted_partitions_.end()) {
+        return;
+    }
+    SortedPartition res = sorted_partitions_.at({attr_list[0]});
+    for (size_t i = 1; i < attr_list.size(); ++i) {
+        res.Intersect(sorted_partitions_.at({attr_list[i]}));
+    }
+    sorted_partitions_.emplace(attr_list, res);
+}
+
+void Order::ComputeDependencies(ListLattice::LatticeLevel const& lattice_level) {
```

Comment:
It would be great if you could decompose this function along with `Order::CreateSingletonSortedPartitions()` (and maybe `Order::UpdateCandidateSets()`), it's really hard to read them